pallets/name-service/src/lib.rs

```diff
@@ -0,0 +1,405 @@
+        #[pallet::weight(T::WeightInfo::remove_register())]
+        #[transactional]
+        pub fn remove_register(
+            origin: OriginFor<T>,
+            username: UserName,
+            registrant: ZkAddressType,
+        ) -> DispatchResult {
+            let _who = ensure_signed(origin)?;
+
+            Self::try_remove_register(username, registrant)?;
+
+            Ok(())
+        }
+    }
+}
+
+impl<T: Config> Pallet<T> {
+    /// Returns the account ID of this pallet.
+    #[inline]
+    pub fn account_id() -> T::AccountId {
+        T::PalletId::get().into_account_truncating()
+    }
+
+    /// Queue username for regiser
+    fn do_register(username: &UserName, registrant: ZkAddressType) -> DispatchResult {
+        // Username checks
+        username_validation(username).ok_or(Error::<T>::InvalidUsernameFormat)?;
+
+        let (hash_user, hash_address) = (
+            T::Hashing::hash_of(username),
+            T::Hashing::hash_of(&registrant),
+        );
+
+        // Check if already Pending Register
+        ensure!(
+            !PendingRegister::<T>::contains_key(hash_user),
+            Error::<T>::AlreadyPendingRegister
+        );
+
+        // Check if already registered
+        ensure!(
+            !UsernameRecords::<T>::contains_key(username),
+            Error::<T>::NameAlreadyRegistered
+        );
+
+        PendingRegister::<T>::insert(
+            hash_user,
+            (
+                hash_address,
+                frame_system::Pallet::<T>::block_number()
+                    .saturating_add(T::RegisterWaitingPeriod::get()),
+            ),
+        );
+
+        Self::deposit_event(Event::NameQueuedForRegister);
+        Ok(())
+    }
+
+    /// Finish Register after block time has passed
+    fn do_accept_register(username: &UserName, registrant: ZkAddressType) -> DispatchResult {
+        // Username checks
+        username_validation(username).ok_or(Error::<T>::InvalidUsernameFormat)?;
+
+        let (hash_user, hash_address) = (
+            T::Hashing::hash_of(username),
+            T::Hashing::hash_of(&registrant),
+        );
+
+        // check if block number has been passed
+        let pending_register_data =
+            PendingRegister::<T>::get(hash_user).ok_or(Error::<T>::UsernameNotFound)?;
+        ensure!(
+            frame_system::Pallet::<T>::block_number() > pending_register_data.1,
+            Error::<T>::RegisterTimeNotReached
+        );
+
+        ensure!(
+            pending_register_data.0 == hash_address,
+            Error::<T>::NotOwned
+        );
+
+        // Move from pending into records
+        PendingRegister::<T>::remove(hash_user);
+        UsernameRecords::<T>::insert(username, registrant);
+
+        Self::deposit_event(Event::NameRegistered);
+        Ok(())
+    }
+
+    /// Set primary name if register and owned
+    fn try_set_primary_name(username: UserName, registrant: ZkAddressType) -> DispatchResult {
+        //check if name is registered
```

Comment:
we should validate the username length here as well before searching

Reply:
Yeah that should limit storage operations